### ISSUE TITLE
Update DOM when changing template in reactive form fails

### ISSUE THREAD

**Issue by natcohen** [angular/angular#14376]
**I'm submitting a ...**  (check one with "x")
```
[ x ] bug report => search github for a similar issue or PR before submitting
[   ] feature request
[   ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

After patching a form of type `FormArray` (`form.patchValue(myArray);`) with `FormGroup` items,  `[formGroup]='form'` in my html throws `control.registerOnChange is not a function`. I patch it when I initialize my app and everything works fine but when I change the form with another `FormArray` (I tried with the same array), I'm getting the error. 

**Expected behavior**

It should not throw any error...

**Minimal reproduction of the problem with instructions**

https://plnkr.co/edit/dJ1KyDLFDOOa1FUfWu5d?p=preview

Click on new template and check the console

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

NodeJS 7.4.0 running on Windows 10

* **Angular version:** 2.4.1

* **Browser:** Chrome 55 64bit

* **Language:** Typescript

**Comment by DzmitryShylovich** [comment]
Duplicate of https://github.com/angular/angular/issues/13788

**Comment by natcohen** [comment]
@DzmitryShylovich Yes nobody gives any attention to this issue... I'm just asking for a bug label (so the bug doesn't get lost) and nobody from the team answers!

**Comment by DzmitryShylovich** [comment]
The team is busy on more priority things because they need to release 4.0-rc0 next week.

**Comment by natcohen** [comment]
@DzmitryShylovich Thanks for the quick response... the problem is that I'm stuck with this bug and can't deliver the app to my client (more than 1 month that I've posted the initial issue and more than a month of delay).

I have tried to look into the code to see if I could help without success.

What can I do to accelerate the process? How can I make sure my issue doesn't get lost and will be resolved in priority (I consider it a priority because everything crashes)?

Thanks again

**Comment by DzmitryShylovich** [comment]
@natcohen nobody is working on forms right now. So I would recommend to reconsider your approach.

**Comment by natcohen** [comment]
@DzmitryShylovich that's a pretty tough statement... After months of development, reconsidering the approach would take additional months!
I don't get the point of abandoning v2 while these bugs will be in v4. I expected the team to work on blocking bugs in parallel to v4. Blocking bugs should be THE priority.

**Comment by aitboudad** [comment]
@natcohen for a workaround try to move the form into a separate component and try to create it dynamically ( dispose the old one / create new instance ) in `changeTemplate` by using `ComponentFactoryResolver`

**Comment by natcohen** [comment]
@aitboudad Thanks for the workaround...will try that

@DzmitryShylovich Any chance this gets resolved in v4 RC?

**Comment by natcohen** [comment]
This bug has not been fixed for v4... can someone have a look?

**Comment by natcohen** [comment]
@mhevery just read you are now in the form team... any chance you can have a look at this issue?

**Comment by lazaridis-com** [comment]
> What can I do to accelerate the process? How can I make sure my issue doesn't get lost and will be resolved in priority (I consider it a priority because everything crashes)?

@natcohen , a) try the suggested workarounds. b) Find a workaround. c) Fix it yourself. d) hire a developer to fix it for you.

@DzmitryShylovich , now **is** this a bug, or not? @natcohen communicates like a commercial customer with a support-contract, which possibly annoys the team - I understand this. But within 3 months, you should really have categorized this issue ( #13788 has just gotten a `comp: forms` label, but nothing else).

**Comment by natcohen** [comment]
@lazaridis-com
a) I have tried many things to fix that bug on my hand... unfortunately I can't afford to spend more time
b) it took me DAYS to find this bug, it wasn't obvious to find in my case! So no, I'm not acting/communicating like a commercial customer but like a frustrated user... the Angular team is trying to attract as many users as they can and when you're stuck with it, they don't offer the support you need. It is frustrating especially when you work few days on the problem and wait 3 months to not even get a clear answer!
c) thanks for trying to help!

**Comment by lazaridis-com** [comment]
@natcohen , I understand your pain. I got even frustrated by just following this issue. Patience, possibly one team-member sees this as a matter of **professional pride** and reacts.

**Comment by ghetolay** [comment]
I don't think this is bug, you can't change a form, you build it only once at init. 

What are you trying to do here ?

**Comment by natcohen** [comment]
@ghetolay Well... I need to change the form based on user interactions. If I can't change a form then it should be specified in the documentation!

What I'm trying to do is a survey platform (like surveymonkey) which supports multiple pages. Each time the user click on the next button, it goes to the next page and change the form.

What I might have to do is to remove all the controls and build again...but why couldn't I just change the form? This should definitely be supported!

**Comment by ghetolay** [comment]
@natcohen 

Ok then looks like you'll have to do it manually using `push()` and `removeAt()` and not `addControl() / removeControl()` since you're using a `FormArray` as root.

**Comment by lazaridis-com** [comment]
Lets see, I'll play a little bit "Product Manager" here

@ghetolay thinks:
> I don't think this is a bug, you can't change a form, you build it only once at init.

@natcohen counters:
> What I might have to do is to remove all the controls and build again...but why couldn't I just change the form? This should definitely be supported!

And I want to ask:

When using plain ECMAScript within a browser, can I change a form at runtime?

If the answer is "no", then Angular `should` provide it (to be stronger than ECMAScript)

If the answer is "yes", then Angular `must` provide it, too, otherwise Angular is weaker than plain ECMAScript, which would be a shooting point ("What? Angular? Too restrictive, you even cannot modify a form at run-time.").

As for this issue:

* If this functionality is a `should` requirement, then this issue becomes a `feature-request`
* If this functionality is a `must` requirement, thin this issue becomes a `bug`

(This classification should **never** take 3 months)

Either you want it or not: this is a Google led project, so one expects some level of professionalism here.

Am I wrong somewhere?

Accepting counters!

And bets!

**Comment by ghetolay** [comment]
> you can't change a form, you build it only once at init.

Ok let me rephrase here. What I meant is you can't change the form object defining the form i.e.: you can't do `this.form = ...` after component initiation (after `ngOnInit()` hook).

But you do can change a form or define it at run-time.

**Comment by natcohen** [comment]
@ghetolay Please explain why we "shouldn't"? I can think of many cases where this is necessary!

**Comment by ghetolay** [comment]
Ok the more I read what I wrote and the more I was doubting it.

So I checked the source and `FormGroupDirective` indeed react to new form reference and do some `cleanupControl()` / `setupControl()` so it seems you can change the form reference and it'll adapt.

## Problem
the error occurs because for the same path `[0, 'answer']` you went from a `FormControl` to a `FormArray`. And so `setupControl()` gets called with that `FormArray` while that function is expecting a `FormControl` as argument so it throws at some point.

So I don't know if it's expected or not or if it's a bug. 
My guess is that path should be unique so that's why having 2 different controls on same path may be confusing `FormGroupDirective`.

## Possible fix
A quick fix I can see would be to change those lines https://github.com/angular/angular/blob/3e51a1998304ab6a15e5bea6bc66e7a8c636a8ad/packages/forms/src/directives/reactive_directives/form_group_directive.ts#L154-L155
with 
```ts
if (newCtrl && newCtrl instanceof FromControl ) {
  setUpControl(newCtrl, dir);
  dir._control = newCtrl;
}
else
  dir._control = null;
```

But I don't know the impact this could have. Maybe @DzmitryShylovich can take it up from here.

## Workaround

My workaround is to use a `FormGroup` as the global form instead of a `FormArray` and to use a different name each time so path are always unique : 
https://plnkr.co/edit/37rCHuJkuyBuV6UjegOt?p=preview

I also think this could make sense for your case since you said you were going from page to page so you could use the page number as the group name for example.

**Comment by natcohen** [comment]
@ghetolay Thanks for your help! For the workaround, it's not doable...I've already built the app and I can't afford to change everything. Though an alternative would be to simply remove each control one by one before re-assigning the form.

For the fix, would you like to do a PR? I can do it but it is your work and you deserve the credit...

**Comment by ghetolay** [comment]
PR open here: https://github.com/angular/angular/pull/15812

**Comment by natcohen** [comment]
@tbosch & @DzmitryShylovich 9 months now... any chance someone can look at this bug and fix it for v5? This is a real bug and it doesn't even have the bug label.

**Comment by natcohen** [comment]
@tbosch, @DzmitryShylovich, @jasonaden This is becoming a joke! Do we need 2 years to fix a bug with a proposed solution?

This whole marketing behind Angular should be followed by some support. You can't provide your framework and, because it's free, leaving users with such bugs for so long. It took me few days of work to find the bug and you don't even answer to give a status.
